File: CSharpDemo/AsyncAwaitWebApi/Controllers/AsyncController.cs
using Microsoft.AspNetCore.Mvc;
using AsyncAwaitWebApi.Services;
using System.Diagnostics;

namespace AsyncAwaitWebApi.Controllers;

[ApiController]
[Route("api/async")]
public class AsyncController : ControllerBase
{
    private readonly DatabaseService _databaseService;
    private readonly ExternalApiService _externalApiService;
    private readonly MetricsService _metricsService;

    public AsyncController(
        DatabaseService databaseService,
        ExternalApiService externalApiService,
        MetricsService metricsService)
    {
        _databaseService = databaseService;
        _externalApiService = externalApiService;
        _metricsService = metricsService;
    }

    /// <summary>
    /// ASYNC: Fetches users from database asynchronously
    /// Thread is RELEASED while waiting for I/O
    /// </summary>
    [HttpGet("users")]
    public async Task<IActionResult> GetUsersAsync()
    {
        var sw = Stopwatch.StartNew();

        // GOOD: Thread is released during await, can handle other requests
        var users = await _databaseService.GetUsersAsync();

        sw.Stop();
        _metricsService.RecordRequest("async/users", sw.ElapsedMilliseconds);

        return Ok(new
        {
            Method = "ASYNC",
            Users = users,
            ResponseTimeMs = sw.ElapsedMilliseconds,
            ThreadId = Environment.CurrentManagedThreadId,
            Message = "Thread was RELEASED during I/O wait"
        });
    }

    /// <summary>
    /// ASYNC: Fetches order details with multiple parallel I/O operations
    /// Demonstrates parallel async calls
    /// </summary>
    [HttpGet("orders/{id}")]
    public async Task<IActionResult> GetOrderAsync(int id)
    {
        var sw = Stopwatch.StartNew();

        // GOOD: Multiple I/O operations run in PARALLEL
        var orderDetails = await _databaseService.GetOrderDetailsAsync(id);
        var enrichedData = await _externalApiService.GetEnrichedDataAsync($"user_{id}");

        sw.Stop();
        _metricsService.RecordRequest("async/orders", sw.ElapsedMilliseconds);

        return Ok(new
        {
            Method = "ASYNC",
            OrderDetails = orderDetails,
            EnrichedData = enrichedData,
            ResponseTimeMs = sw.ElapsedMilliseconds,
            ThreadId = Environment.CurrentManagedThreadId,
            Message = "Multiple I/O operations ran in PARALLEL, thread was RELEASED"
        });
    }

    /// <summary>
    /// ASYNC: Complex scenario with multiple I/O operations
    /// </summary>
    [HttpGet("complex")]
    public async Task<IActionResult> ComplexOperationAsync()
    {
        var sw = Stopwatch.StartNew();

        // All these I/O operations run in parallel
        var usersTask = _databaseService.GetUsersAsync();
        var weatherTask = _externalApiService.GetWeatherAsync("New York");
        var orderTask = _databaseService.GetOrderDetailsAsync(123);

        // Wait for all to complete - thread is released
        await Task.WhenAll(usersTask, weatherTask, orderTask);

        sw.Stop();
        _metricsService.RecordRequest("async/complex", sw.ElapsedMilliseconds);

        return Ok(new
        {
            Method = "ASYNC",
            Users = usersTask.Result,
            Weather = weatherTask.Result,
            Order = orderTask.Result,
            ResponseTimeMs = sw.ElapsedMilliseconds,
            ThreadId = Environment.CurrentManagedThreadId,
            Message = "All I/O operations ran in PARALLEL"
        });
    }
}

File: NodeDemo/src/controllers/asyncController.ts
import { Request, Response } from 'express';
import { DatabaseService } from '../services/databaseService';
import { ExternalApiService } from '../services/externalApiService';
import { MetricsService } from '../services/metricsService';

export class AsyncController {
    constructor(
        private dbService: DatabaseService,
        private apiService: ExternalApiService,
        private metricsService: MetricsService
    ) {}

    /**
     * ASYNC: Fetches users from database asynchronously
     * Event loop is FREE during I/O wait - can handle other requests
     */
    async getUsers(req: Request, res: Response): Promise<void> {
        const start = Date.now();

        try {
            // GOOD: Event loop is free during await
            const users = await this.dbService.getUsersAsync();

            const responseTime = Date.now() - start;
            this.metricsService.recordRequest('async/users', responseTime);

            res.json({
                method: 'ASYNC',
                users,
                responseTimeMs: responseTime,
                message: 'Event loop was FREE during I/O wait'
            });
        } catch (error) {
            res.status(500).json({ error: 'Internal server error' });
        }
    }

    /**
     * ASYNC: Fetches order details with multiple parallel I/O operations
     * Demonstrates parallel async calls
     */
    async getOrder(req: Request, res: Response): Promise<void> {
        const start = Date.now();
        const orderId = parseInt(req.params.id as string);

        try {
            // GOOD: Multiple I/O operations run in PARALLEL
            const [orderDetails, enrichedData] = await Promise.all([
                this.dbService.getOrderDetailsAsync(orderId),
                this.apiService.getEnrichedDataAsync(`user_${orderId}`)
            ]);

            const responseTime = Date.now() - start;
            this.metricsService.recordRequest('async/orders', responseTime);

            res.json({
                method: 'ASYNC',
                orderDetails,
                enrichedData,
                responseTimeMs: responseTime,
                message: 'Multiple I/O operations ran in PARALLEL, event loop was FREE'
            });
        } catch (error) {
            res.status(500).json({ error: 'Internal server error' });
        }
    }

    /**
     * ASYNC: Complex scenario with multiple parallel I/O operations
     */
    async complexOperation(req: Request, res: Response): Promise<void> {
        const start = Date.now();

        try {
            // All these I/O operations run in parallel
            const [users, weather, order] = await Promise.all([
                this.dbService.getUsersAsync(),
                this.apiService.getWeatherAsync('New York'),
                this.dbService.getOrderDetailsAsync(123)
            ]);

            const responseTime = Date.now() - start;
            this.metricsService.recordRequest('async/complex', responseTime);

            res.json({
                method: 'ASYNC',
                users,
                weather,
                order,
                responseTimeMs: responseTime,
                message: 'All I/O operations ran in PARALLEL'
            });
        } catch (error) {
            res.status(500).json({ error: 'Internal server error' });
        }
    }
}
